### PR TITLE
Reduce some allocations in parser

### PR DESCRIPTION
**Pull request by peterebden**
Attempts to size scopes a little more accurately, and parses some expressions in-place rather than creating a new struct and copying.

### CODE REVIEW

File: src/parse/asp/grammar_parse.go [hunk 1]
```diff
@@ -430,41 +428,47 @@ func (p *parser) parseUnconditionalExpressionInPlace(e *Expression) {
 	}
 }
 
-func concatStrings(lhs *ValueExpression, rhs *ValueExpression) *ValueExpression {
-	// If they're both fStrngs
+func concatStrings(lhs *ValueExpression, rhs *ValueExpression) {
```
Comment (Tatskaari) [inline src/parse/asp/grammar_parse.go:431]: This is doing the same thing just isn't returning a value to be clearer that it's mutating the parameter?

**Review by Tatskaari — APPROVED** [review]
I think this looks sensible. Does it make a big difference?

**Comment by peterebden** [comment]
> I think this looks sensible. Does it make a big difference?

Nope, it only makes a small difference. Just trying to collect a few of these for some relatively low-hanging fruit on reducing GC pressure.

**Comment by peterebden** [comment]
OK so I've written a benchmark for this too:
```
BenchmarkParseFile
BenchmarkParseFile-12                260           4462050 ns/op         1632444 B/op      16476 allocs/op
```
vs.
```
BenchmarkParseFile
BenchmarkParseFile-12                274           4039847 ns/op         1571474 B/op      16336 allocs/op
```
(I have run both a few times and the numbers are pretty consistent)
So it's saving about 1% of allocations which isn't massive but easy enough; I've also dropped the parser changes so it's just scope sizing now.

**Pull request merged**